nft_ingester/src/backfiller.rs

```diff
@@ -0,0 +1,702 @@
+                                    }
+                                }
+                                Err(err) => {
+                                    println!("Failed to fetch and plug gaps for {tree_str}");
+                                    println!("{err}");
+                                    self.sleep_and_increase_delay().await;
+                                }
+                            }
+                        }
+                    }
+                }
+                Err(err) => {
+                    // Print error but keep trying.
+                    println!("Could not get trees to backfill from db: {err}");
+                    self.sleep_and_increase_delay().await;
+                }
+            }
+        }
+    }
+
+    async fn sleep_and_increase_delay(&mut self) {
+        sleep(Duration::from_millis(self.failure_delay)).await;
+
+        // Increase failure delay up to `MAX_FAILURE_DELAY_MS`.
+        self.failure_delay = self.failure_delay.saturating_mul(2);
+        self.failure_delay = if self.failure_delay > MAX_FAILURE_DELAY_MS {
+            MAX_FAILURE_DELAY_MS
+        } else {
+            self.failure_delay
+        };
+    }
+
+    fn reset_delay(&mut self) {
+        self.failure_delay = INITIAL_FAILURE_DELAY;
+    }
+
+    async fn get_trees_to_backfill(&self) -> Result<Vec<BackfillTree>, DbErr> {
+        // Get trees with the `force_chk` flag set.
+        let force_chk_trees = UniqueTree::find_by_statement(Statement::from_sql_and_values(
+            DbBackend::Postgres,
+            "SELECT DISTINCT backfill_items.tree FROM backfill_items\n\
+            WHERE backfill_items.force_chk = TRUE",
+            vec![],
+        ))
+        .all(&self.db)
+        .await?;
+
+        // Convert this Vec of `UniqueTree` to a Vec of `BackfillTree` (which contain extra info).
+        let mut trees: Vec<BackfillTree> = force_chk_trees
+            .into_iter()
+            .map(|tree| BackfillTree::new(tree, true))
+            .collect();
+
+        // Get trees with multiple rows from `backfill_items` table.
+        let multi_row_trees = UniqueTree::find_by_statement(Statement::from_sql_and_values(
+            DbBackend::Postgres,
+            "SELECT backfill_items.tree FROM backfill_items\n\
+            GROUP BY backfill_items.tree\n\
+            HAVING COUNT(*) > 1",
+            vec![],
+        ))
+        .all(&self.db)
+        .await?;
+
+        // Convert this Vec of `UniqueTree` to a Vec of `BackfillTree` (which contain extra info).
+        let mut multi_row_trees: Vec<BackfillTree> = multi_row_trees
+            .into_iter()
+            .map(|tree| BackfillTree::new(tree, false))
+            .collect();
+
+        trees.append(&mut multi_row_trees);
+
+        Ok(trees)
+    }
+
+    async fn backfill_tree_from_seq_1(&self, tree: &[u8]) -> Result<Option<i64>, IngesterError> {
+        //TODO implement gap filler that gap fills from sequence number 1.
```

Comment:
Yes I'm not quite sure how to find the block for sequence number 1 of a non-existent tree yet.  I'm sure it can be done, but I kind of wanted to punt on this functionality for the next phase which may involve looking for missing transactions rather than missing sequence numbers.